Add InlinedCallStack class.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29921 Add a pass that reconstructs scopes for nodes from their inline callstacks.
* #29911 Show full call stack in TorchScript exception even when calls were inlined.
* #27922 Add logging to inliner.
* **#27921 Add InlinedCallStack class.**
* #27920 Make inlineCallTo to take Function instead of Graph as the callee argument.
* #27919 Add a variant of insertGraph that fills values map.

InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

Commit:
Add CallStack class.

Callstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

ZolotukhinM commented:
This is a resubmitted PR, the old one (#21939) got bit rot.

Commit:
Update on "Add CallStack class."

Add CallStack class.

Callstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

zdevito commented on torch/csrc/jit/scope.cpp:
```diff
@@ -77,5 +78,50 @@ std::string Scope::namesFromRoot(const std::string& separator) const {
+  return caller_;
+}
+
+std::vector<CallStackEntry> CallStack::asVector() {
```
We've called this `vec()` in other contexts. I.e. turning an ArrayRef into a vector.

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -1652,10 +1669,24 @@ at::ArrayRef<Value*> createTupleUnpack(Value* v) {
+    Node* orig_node = kv.first->node();
+    Node* new_node = kv.second->node();
+    if (updated_nodes.insert(new_node).second) {
+      new_node->insertCallStackEntry(callee, to_replace->sourceRange());
```
I believe insertGraph preserves source ranges. But it looks like it does not preserve callstacks as well? That seems inconsistent. My expectation would be that it would not drop part of the information about a node.  It might make more sense for insertGraph to take an optional `callee_function` and have it do the bookkeeping to update the graph, otherwise it would copy the callstack information as is. The current algorithm has the Print issue as documented above, and is nondeterministic in the order it processes nodes. This isn't wrong, but will make debugging errors around this code harder.

ZolotukhinM replied:
I changed how we update the callstacks in insertGraph : now when we copy a node we copy its callstack as well and then here we just need to add one more entry to it (that also allowed me to delete "appendCallstack.." function).

zdevito reviewed (commented):
This looks pretty good. I think it would be good to document what the expected callstack information looks like for a Node if it were (1) not inlined, (2) inlined once, and (3) inlined twice. Right now I do not know whether Function* of the current function shows up in the callstack of every node (I am guessing no, but am unsure), and the order of the nodes in the callstack list (top to bottom or bottom to top).

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
rebase on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

ZolotukhinM reviewed (commented):
I updated the PR, PTAL!

zdevito commented on torch/csrc/jit/scope.cpp:
```diff
@@ -77,5 +78,50 @@ std::string Scope::namesFromRoot(const std::string& separator) const {
+  c10::optional<InlinedCallStackPtr> current = intrusive_from_this();
+  while (current) {
+    r.emplace_back(std::make_pair((*current)->fn_, (*current)->source_range_));
+    current = (*current)->caller_;
```
Something is confusing to me here. The _last_ thing in this list is the inner-most function, right? The docs say that and the tests also test that. But the inner-most thing cannot be the "caller" of anything. Nevertheless, if it is the last thing in this loop, current->caller_ must at some point point to this inner-most thing. I believe that the `caller_` `callee_` names are swapped in the Trie. My follow up thought is that I don't understand how the Trie is memoizing anything. The inner most link is the most inlined node, but this node will only appear once. Is it actually caching anything?

ZolotukhinM replied:
You were correct, the caching didn't work properly. I fixed that and added a test to verify that we're actually sharing callstack-entries among different nodes when it's possible. Please take another look!

zdevito reviewed (commented):
Thanks for adding the comments about what the stack represents. See the inline comment about naming and use of Trie.

ZolotukhinM commented on torch/csrc/jit/ir.cpp:
```diff
@@ -1651,10 +1661,22 @@ at::ArrayRef<Value*> createTupleUnpack(Value* v) {
+
+  // TODO: We might need to use nodes_map instead of value_map. Otherwise, we
+  // are missing nodes without outputs (e.g. prim::Print).
+  std::unordered_set<Node*> updated_nodes;
```
I have mixed feeling about this. On one hand, we indeed would move the update code inside the `insertGraph` which will fix the TODO and would allow us to get rid of the `value_map` variant. On the other hand, we would introduce two new arguments (`c10::optional<Function*> callee` and `c10::optional<SourceRange> sr`) which will have overlapping meaning with existing arguments (i.e. "Must `callee->graph()` be the same as `graph`? Should we check that? Should we fail if not?"). It would be ok if we could remove the `graph` argument and replace it with `callee`, but there are some use-cases which insert graph and not the function, so it wouldn't work for them.

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 7186fa5:
* 0/8 recognized as flaky
* 8/8 broken upstream. You may want to rebase on the [latest viable branch](https://dr.pytorch.org/master-viable-commits.html).
* 0/8 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=7186fa5a4ec0581bdb19fc9bcf7b77505c9febf8).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
rebase on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

zdevito reviewed (commented):
Discussed in person: We can simplify by putting the the caching logic directly in inlineCallTo, since all the repeated node creations come into existence during this step as prepends to the link list of call stack entries. This allows us to have just a single CallStackEntry type, putting a `std::vector<CallStackEntry> Node::callstack();`

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

ZolotukhinM commented:
I reimplemented it according to our previous discussion and I think it looks much cleaner now. @zdevito, please take another look!

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head

Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -1712,10 +1714,43 @@ at::ArrayRef<Value*> createTupleUnpack(Value* v) {
+
+    auto new_node_cs = new_node->callstack();
+
+    InlinedCallStack* raw_callstack_ptr =
```
Isn't this just equivalent to `raw_callstack_ptr = new_node_cs->get()`?

ZolotukhinM replied:
We have two dereferences there: first from Optional, second from intrusive_ptr to a raw pointer. The selection statement is for resolving the optional, I believe we cannot drop it.

zdevito reviewed (approved):
Cool, this looks great now!

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
Update on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]

Commit:
rebase on "Add InlinedCallStack class."


InlinedCallstack serves a similar purpose to Scope, but instead of storing
string names of the functions it stores pointer to Function objects
themselves. Currently, scopes are used in tracing and callstacks are
used in scripting - hopefully I would be able to merge them in future.

gh-metadata: pytorch pytorch 27921 gh/ZolotukhinM/139/head


Differential Revision: [D17914132](https://our.internmc.facebook.com/intern/diff/D17914132)

[ghstack-poisoned]
